### PR TITLE
[FSDP2] Added `shard_placement_fn` arg

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136221

## TL;DR
This PR adds a `shard_placement_fn: Optional[Callable[[nn.Parameter], Optional[Shard]]` arg to `fully_shard` that allows users to specify FSDP sharding on a nonzero tensor dim. If doing so, then the tensor dim size must be divisible by the FSDP shard world size.

```
# Example:
def shard_placement_fn(param: nn.Parameter) -> Optional[Shard]:
    largest_dim = largest_dim_size = -1
    for dim, dim_size in enumerate(param.shape):
        if dim_size > largest_dim_size:
            largest_dim = dim
            largest_dim_size = dim_size
    return Shard(largest_dim)

fully_shard(module, shard_placement_fn=shard_placement_fn)
```

## Follow-Ups
- **Copy kernels:** For all-gather copy-out, we currently copy-out to temporaries and then chunk-dim-0 -> cat-shard-dim, incurring an extra copy for parameters sharded on nonzero tensor dim. Similarly, for reduce-scatter copy-in, we currently chunk-shard-dim -> cat-dim-0, incurring an extra copy for gradients sharded on nonzero tensor dim. @yifuwang  has ideas for adding additional split size args to the copy ops that allows fusing these extra copies into the existing all-gather copy-out and reduce-scatter copy-in.

cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D62964657](https://our.internmc.facebook.com/intern/diff/D62964657)

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_init.py [hunk 1]
```diff
@@ -993,5 +993,105 @@ def test_hsdp_broadcast_across_replicas(self):
...
+                    largest_dim_size = dim_size
+            assert largest_dim >= 0, f"{param.shape}"
+            return Shard(largest_dim)
+
+        for layer in model.layers:
+            fully_shard(layer, shard_placement_fn=shard_placement_fn)
+        fully_shard(model, shard_placement_fn=shard_placement_fn)
+
+        for param, ref_param in zip(model.parameters(), ref_model.parameters()):
+            full_param = param.full_tensor()
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_init.py:1029]: if I print params here, some params would be shard(1), and some params are shard(0)? 

We expect state dict to be composable with this PR because DTensor full_tensor/redistribute will handle placement = shard(0) / shard(1) gracefully?

Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_init.py:1029]: > if I print params here, some params would be shard(1), and some params are shard(0)?

Yep!

> We expect state dict to be composable with this PR because DTensor full_tensor/redistribute will handle placement = shard(0) / shard(1) gracefully?

That is my understanding.

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 2]
```diff
@@ -309,6 +357,20 @@ def foreach_reduce(
         reduce_scatter_group, all_reduce_group, reduce_dtype
     )
     world_size = reduce_scatter_group.size()
+    for i, (fsdp_param, unsharded_grad) in enumerate(zip(fsdp_params, unsharded_grads)):
+        shard_dim = fsdp_param.fsdp_placement.dim
+        if shard_dim != 0:
+            if unsharded_grad.size(shard_dim) % world_size != 0:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:363]: For now, I could not come up with a way to avoid this extra copy for unsharded gradients that require padding when we are not sharding on dim-0.

cc: @yifuwang in case you have any ideas
(more generally, if you look through the changes, there are a few places where we did 'tricks' like `torch.as_strided` with contiguous strides to reduce CPU overhead, but those fell apart once sharding on non-zero dim)

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62964657).

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 3]
```diff
@@ -320,48 +330,66 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
+            raise AssertionError(
+                f"Shard dim {shard_dim} is invalid for {param_data.ndim}D tensor: {param.shape}"
+            )
         self._orig_size = param_data.size()
         self._contiguous_orig_stride = make_contiguous_strides_for(self._orig_size)
         shard_rank = self.mesh_info.shard_mesh_rank
         shard_world_size = self.mesh_info.shard_mesh_size
-        chunks = _chunk_with_empty(param_data, shard_world_size, dim=0)
+        chunks = _chunk_with_empty(param_data, shard_world_size, dim=shard_dim)
+        self._shard_chunk_sizes = [c.size() for c in chunks]
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:357]: to-do: evaluate if there is a cleaner way

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 4]
```diff
@@ -320,48 +330,66 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
+            sharded_param, param_data.size(), dim=shard_dim
+        )
         self.contiguous_sharded_stride = make_contiguous_strides_for(self.sharded_size)
         padded_sharded_size = chunks[0].size()  # 0th always padded
-        padded_sharded_param = param_data.new_zeros(padded_sharded_size)
-        self.padded_sharded_param_size = padded_sharded_param.size()
-        if sharded_param.numel() > 0:
-            padded_sharded_param[: sharded_param.size(0)].copy_(sharded_param)
+        self.padded_sharded_param_size = padded_sharded_size
+        if shard_dim == 0:
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:365]: should we branch on `shard_dim == 0` or `shard_dim == 0 or not requires_padding`

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 5]
```diff
@@ -470,12 +500,23 @@ def init_unsharded_param(self):
...
+            )
+        else:
+            padded_unsharded_size = list(self.padded_sharded_param_size)
+            world_size = (
+                unsharded_tensor.numel() // self.padded_sharded_param_size.numel()
+            )
+            padded_unsharded_size[shard_dim] *= world_size
+            unsharded_param = unsharded_tensor.view(padded_unsharded_size).narrow(
+                dim=shard_dim, start=0, length=self._orig_size[shard_dim]
+            )
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:519]: to-do: this may be non-contiguous --> we need to copy-out again?

**Comment by awgu** [comment]
To-do:
- Add some state dict test
- (Optional) add some memory test (will use more memory though)
- Test on 8-GPU devgpu in torchtitan and sanity check profiler trace

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 6]
```diff
@@ -320,48 +332,67 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
+        )
         self.contiguous_sharded_stride = make_contiguous_strides_for(self.sharded_size)
         padded_sharded_size = chunks[0].size()  # 0th always padded
+        self.padded_sharded_param_size = padded_sharded_size
+        # Pre-pad the sharded parameter to avoid padding before all-gather
         padded_sharded_param = param_data.new_zeros(padded_sharded_size)
-        self.padded_sharded_param_size = padded_sharded_param.size()
         if sharded_param.numel() > 0:
-            padded_sharded_param[: sharded_param.size(0)].copy_(sharded_param)
+            padded_sharded_param.narrow(
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:377]: I was worried about NF4 but found that `.narrow` and `[:]` both dispatch to `slice.Tensor`. That's good! any ideas on how to find the mapping? I am a litle bit surpised since I found narrow in native_functions.yaml

Comment (weifengpy) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:377]:
```
import torch
from torch.utils._python_dispatch import TorchDispatchMode


class LoggingMode(TorchDispatchMode):
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        # Print the function being dispatched and the arguments
        print(f"Dispatching function: {func.__name__}")
        return func(*args, **kwargs)

# Example usage
data = torch.rand(3, 3, device="cuda")

# Use LoggingMode for the duration of this block
with LoggingMode():
    data.narrow(dim=1, start=0, length=2)
    data[:2]
```

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 7]
```diff
@@ -320,48 +332,67 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
-        self.sharded_param = nn.Parameter(
-            self.to_sharded_dtensor(padded_sharded_param[: sharded_param.size(0)])
-        )
+        if padded_sharded_param.numel() > 0:
+            length = sharded_param.size(shard_dim) if sharded_param.numel() > 0 else 0
+            sharded_param = padded_sharded_param.narrow(
+                dim=shard_dim, start=0, length=length
+            )
+        else:
+            sharded_param = padded_sharded_param.view(self.sharded_size)
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:393]: curious when will we trigger this branch, `padded_sharded_param.numel() == 0`? is this to cover the edge case when `model.parameters()` returns a param with shape `()` ?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 8]
```diff
@@ -313,6 +350,14 @@ def foreach_reduce(
...
     )
     world_size = reduce_scatter_group.size()
+    for i, (fsdp_param, unsharded_grad) in enumerate(zip(fsdp_params, unsharded_grads)):
+        if (shard_dim := fsdp_param.fsdp_placement.dim) == 0:
+            continue
+        assert (
+            unsharded_grad.size(shard_dim) % world_size == 0
+        ), f"Shard({shard_dim}) requires even sharding: {unsharded_grad.size()=} {world_size=}"
+        chunks = torch.chunk(unsharded_grad, world_size, dim=shard_dim)
+        unsharded_grads[i] = torch.cat(chunks, dim=0)
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:360]: are we chunk + cat to make unsharded_grads contiguous? does the `.cat` trigger copies on gpu?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 9]
```diff
@@ -206,11 +206,13 @@ def use_foreach_copy(fsdp_param: FSDPParam) -> bool:
...
-                if fsdp_param.sharded_state == ShardedState.SHARDED
-                else cast(torch.Tensor, fsdp_param._sharded_post_forward_param_data)
-            )
+            if fsdp_param.sharded_state == ShardedState.SHARDED:
+                all_gather_input = fsdp_param._sharded_param_data
+            else:
+                all_gather_input = cast(
+                    torch.Tensor, fsdp_param._sharded_post_forward_param_data
+                )
+            assert all_gather_input.ndim == 1, f"{all_gather_input.shape=}"
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:215]: just a refactor to be able to assert on flat input?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 10]
```diff
@@ -252,34 +254,69 @@ def foreach_all_gather_copy_out(
...
+            device,
+            force_recreate=force_recreate,
+        )
+        if not force_recreate:
             fsdp_param.alloc_all_gather_outputs()
+        param_all_gather_outputs = fsdp_param.all_gather_outputs
+        if fsdp_param.fsdp_placement.dim != 0:
+            # Copy to a temporary and then chunk-cat into the final all-gather
+            # output tensors
+            param_all_gather_outputs = [
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:279]: seems it's overwriting `init_all_gather_outputs`. is it better to move the logic into `init_all_gather_outputs` ? or this is only temporary ?

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py [hunk 11]
```diff
@@ -252,34 +254,69 @@ def foreach_all_gather_copy_out(
...
+            pre_param_size = list(padded_sharded_size)
+            pre_param_size[0] *= world_size
+            chunks = torch.chunk(
+                param_all_gather_output.view(pre_param_size), world_size, dim=0
+            )
+            post_param_size = list(padded_sharded_size)
+            post_param_size[shard_dim] *= world_size
+            cat_out = target_all_gather_output.view(post_param_size)
+            torch.cat(chunks, dim=shard_dim, out=cat_out)
+            torch._C._autograd._unsafe_set_version_counter(
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_collectives.py:316]: is this similar to `with _unsafe_preserve_version_counter(target_all_gather_output)`? we use `_unsafe_set_version_counter` because `param_all_gather_outputs` is defined in `for ... in` and the api is easier?

File: torch/distributed/_composable/fsdp/_fsdp_param.py [hunk 12]
```diff
@@ -286,16 +299,16 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
...
-            split_factor = self._tp_spec.num_shards_map[0]
+            split_factor = self._tp_spec.num_shards_map[shard_dim]
             assert (
                 2 <= self._spmd_mesh.ndim <= 3
             ), f"_spmd_mesh.ndim can only be 2 or 3 but got {self._spmd_mesh.ndim}."
             self._spmd_placements: Tuple[Placement, ...]
             dp_shard_tp_placement = (
                 (
-                    _StridedShard(0, split_factor=split_factor)
+                    _StridedShard(shard_dim, split_factor=split_factor)
```
Comment (weifengpy) [inline torch/distributed/_composable/fsdp/_fsdp_param.py:309]: I never truly understand strided sharding. I guess it's meant for `model.parameters()` outside of fwd/bwd ? mostly for DTensor ops outside of FSDP, like state dict?

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 7 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:393]: This is when the rank only has padding for the sharded parameter. This code was useful when we supported uneven sharding for `Shard(i)` so that the shape is like `[256, 0]` instead of just `[0]` for example. It might not be needed anymore.

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 8 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:360]: We have to chunk -> cat to do a data shuffle. If I have some time, I may try to draw some diagrams to add to the PR description.

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 9 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:215]: yep -- I can revert this change

For uneven `Shard(i)`, I had to add an `elif` here, but I removed it now.

File: torch/distributed/_composable/fsdp/_fsdp_collectives.py (hunk 11 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_collectives.py:316]: yes `with _unsafe_preserve_version_counter(target_all_gather_output)` calls into this `_unsafe_set_version_counter`
since we know we just need to decrement once, I think it is simpler/lower overhead to just call the API directly

File: torch/distributed/_composable/fsdp/_fsdp_param.py (hunk 12 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param.py:309]: yep

**Comment by awgu** [comment]
re-opened in a different PR since the test-config/distributed label seems to persist even after removing
